xapian-core/cluster/cluster.cc

```diff
@@ -0,0 +1,586 @@
+{
+    if (!started) return false;
+    return i == end;
+}
+
+TermIterator::Internal *
+PointTermIterator::skip_to(const string &term)
+{
+    if (i->first == term)
+	return NULL;
+    while (i->first != term)
+	i++;
+    return NULL;
+}
+
+TermIterator
+PointType::termlist_begin() const
+{
+    LOGCALL(API, TermIterator, "PointType::termlist_begin", NO_ARGS);
+    return TermIterator(new PointTermIterator(values));
+}
+
+bool
+PointType::contains(const string &term) const
+{
+    LOGCALL(API, bool, "PointType::contains", term);
+    return values.find(term) != values.end();
+}
+
+double
+PointType::get_value(const string &term) const
+{
+    LOGCALL(API, double, "Point::get_value", term);
+    unordered_map<string, double>::const_iterator it = values.find(term);
+    return (it == values.end()) ? 0.0 : it->second;
+}
+
+double
+PointType::get_magnitude() const {
+    LOGCALL(API, double, "PointType::get_magnitude", NO_ARGS);
+    return magnitude;
+}
+
+void
+PointType::add_value(const string &term, double value)
+{
+    LOGCALL_VOID(API, "PointType::add_value", term | value);
+    unordered_map<string, double>::iterator it;
+    it = values.find(term);
+    if (it != values.end())
+	it->second += value;
+    else
+	values[term] = value;
+}
+
+void
+PointType::set_value(const string &term, double value)
+{
+    LOGCALL_VOID(API, "PointType::set_value", term | value);
+    values[term] = value;
+}
+
+int
+PointType::termlist_size() const
+{
+    LOGCALL(API, int, "PointType::termlist_size", NO_ARGS);
+    return values.size();
+}
+
+Document
+Point::get_document() const
+{
+    LOGCALL(API, Document, "Point::get_document", NO_ARGS);
+    return doc;
+}
+
+void
+Point::initialize(const TermListGroup &tlg, const Document &doc_)
+{
+    LOGCALL_VOID(API, "Point::initialize", tlg | doc_);
+    doccount size = tlg.get_doccount();
+    doc = doc_;
+    for (TermIterator it = doc.termlist_begin(); it != doc.termlist_end(); ++it) {
+	doccount wdf = it.get_wdf();
+	string term = *it;
+	double termfreq = tlg.get_termfreq(term);
+	/** If term is not a stemmed term or the term indexes only one
+	 *  one document, do not compute TF-IDF scores for them
+	 */
+	if (wdf < 1 || term[0] != 'Z' || termfreq <= 1)
+	    continue;
+	double tf = 1 + log((double)wdf);
+	double idf = log(size / termfreq);
+	double wt = tf * idf;
+	values[term] = wt;
+	magnitude += wt * wt;
+    }
+}
+
+Centroid::Centroid(Point &p) {
+    LOGCALL_CTOR(API, "Centroid", p);
+    for (TermIterator it = p.termlist_begin(); it != p.termlist_end(); ++it)
+	values[*it] = p.get_value(*it);
+    magnitude = p.get_magnitude();
+}
+
+void
+Centroid::divide(double num)
+{
+    LOGCALL_VOID(API, "Centroid::divide", num);
+    unordered_map<string, double>::iterator it;
+    for (it = values.begin(); it != values.end(); ++it)
+	it->second /= num;
+}
+
+void
+Centroid::clear()
+{
+    LOGCALL_VOID(API, "Centroid::clear", NO_ARGS);
+    values.clear();
+}
+
+void
+Centroid::recalc_magnitude()
```

Comment:
I guess the lazy approach shouldn't fit here currently because we need the magnitude of the centroids while assigning Point objects to clusters. For every point, we need to calculate the distance between the point and the cluster. So calculating the new magnitude of the centroid at the end of every iteration saves some time. (Mainly because centroids are way more dense than points)
The lazy approach makes more sense in elkans-kmeans, since we don't calculate all point-centroid distances. But I guess the number of times we calculate the distances will be more than 'rarely'.